src/assignments/benson/advent_2018/task2.clj

```diff
@@ -6,66 +6,85 @@
 
 (def ids (clojure.string/split-lines task2input))
 
-(defn match-2 
-  "returns if id has 2 matching letters"
-  [id]
-  (contains? (set (vals (frequencies id))) 2))
-
-(defn match-3 
-  "returns if id has 3 matching letters"
-  [id]
-  (contains? (set (vals (frequencies id))) 3))
-
-(assert (match-2 "abcdec"))
-(assert (not (match-2 "abcdef")))
-(assert (match-3 "accdec"))
-(assert (not (match-3 "abcdef")))
-(assert (and (match-2 "aacbbb") (match-3 "aacbbb")))
-
-(defn calculate-checksum 
+(defn match-n 
+  "returns if id has N matching letters"
+  [n id]
+  (contains? (set (vals (frequencies id))) n))
+
+
+(assert (match-n 2 "abcdec"))
+(assert (not (match-n 2 "abcdef")))
+(assert (match-n 3 "accdec"))
+(assert (not (match-n 3 "abcdef")))
+(assert (and (match-n 2 "aacbbb") (match-n 3 "aacbbb")))
+
+;; (defn calculate-checksum 
+;; "Returns the checksum calculated by multiplying the count of ID's with 2 matching letters
+;;  and ID's with 3 matching letters"
+;;   [ids]
+;;   (loop [[curr-id & rest-ids] ids
+;;          two-matches 0
+;;          three-matches 0]
+;;     (if curr-id
+;;       (recur rest-ids
+;;              (if (match-2 curr-id) (inc two-matches) two-matches) 
+;;              (if (match-3 curr-id) (inc three-matches) three-matches))
+;;       (* two-matches three-matches))))
+
+(defn match-count 
+  "Returns count of IDS with N matches"
+  [ids n]
+    (reduce (fn [matches id]
+              (if (match-n n id) (inc matches) matches))
+            0
+            ids))
+
+(defn calculate-checksum
   "Returns the checksum calculated by multiplying the count of ID's with 2 matching letters
    and ID's with 3 matching letters"
   [ids]
-  (loop [[curr-id & rest-ids] ids
-         two-matches 0
-         three-matches 0]
-    (if curr-id
-      (recur rest-ids
-             (if (match-2 curr-id) (inc two-matches) two-matches) 
-             (if (match-3 curr-id) (inc three-matches) three-matches))
-      (* two-matches three-matches))))
+  (* (match-count ids 2) (match-count ids 3)))
```

Comment:
So, interestingly, your original solution was more performant, because it traversed the sequence only once.  (Imagine this is some sort of bio-molecular code running on enormous genetic sequences and performance really matters).

Your second solution, however, is more elegant and readable.

Can you see a way to get the best of both worlds?